batch/batch/worker/copy.py

```diff
@@ -0,0 +1,83 @@
+import os
+import re
+import argparse
+import subprocess as sp
+from shlex import quote as shq
+
+from .flock import Flock
+
+
+GCS_PATH_REGEX = re.compile('gs://(?P<bucket>[^/]+)/(?P<path>.+)')
+
+
+def check_call(cmd):
+    print(cmd)
+    return sp.check_call(cmd, shell=True, stderr=sp.STDOUT)
+
+
+wildcards = ('*', '?', '[', ']', '{', '}')
+
+
+def contains_wildcard(c):
+    i = 0
+    n = len(c)
+    while i < n:
+        if i < n - 1 and c[i] == '\\' and c[i + 1] in wildcards:
+            i += 2
+            continue
+        elif c[i] in wildcards:
+            return True
+        i += 1
+    return False
+
+
+def copy(src, dst, user, io_path, cache_path):
+    if contains_wildcard(src):
+        raise NotImplementedError(f'glob wildcards are not allowed in file paths, got source {src}')
+
+    if contains_wildcard(dst):
+        raise NotImplementedError(f'glob wildcards are not allowed in file paths, got destination {dst}')
+
+    if not src.startswith('gs://'):
+        raise NotImplementedError(f'cannot copy from sources that are not GCS paths, got source {src}')
+
+    cache_src = f'{cache_path}/{user}{src[4:]}'
+    escaped_src = re.escape(os.path.basename(src))
+
+    if dst.startswith('/io'):
+        dst = f'/host{io_path}{dst[3:]}'
+
+    match = GCS_PATH_REGEX.fullmatch(src)
+    if not match:
+        raise NotImplementedError(f'source is not a valid GCS path, got source {src}')
+    src_path = match.groupdict()['path']
+    src_tail_dir = src_path.rstrip('/').split('/')[-1]
+    rsync_dst = f'{dst.rstrip("/")}/{src_tail_dir}/'
+
+    with Flock(cache_src):
+        os.makedirs(os.path.dirname(dst), exist_ok=True)
+        os.makedirs(os.path.dirname(cache_src), exist_ok=True)
+        check_call(f'cp -p -R --reflink {shq(cache_src)} {shq(dst)} || true')
+        check_call(f'''
+gsutil -q stat {shq(src)}
+if [ $? = 0 ]; then
```

Comment:
stat returns zero if there is a file whose name ends in `/`. There's an easy fix: if `src` ends in `/`, don't both calling stat and checking the output. Just `check_call(f'''mkdir -p ...)`